### PR TITLE
Add IgnoredInterfaces field to BGPConfiguration

### PR DESCRIPTION
**Pull request by huiyizzz**
## Description

This PR adds a optional field `IgnoredInterfaces` in `BGPConfiguration` resource to ignore interfaces for Direct protocol in BIRD configuration.
## Related issues/PRs

https://github.com/projectcalico/calico/issues/5225
## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
Add ignorable interfaces via the BGPConfiguration API
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

File: confd/pkg/backends/calico/routes_test.go [hunk 1]
```diff
@@ -12,6 +12,9 @@ import (
 
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/client-go/tools/cache"
+
+	apiv3 "github.com/projectcalico/api/pkg/apis/projectcalico/v3"
```
Comment (coutinhop) [inline confd/pkg/backends/calico/routes_test.go:16]: `make check-fmt` in confd/ seems to be failing in this file. Try running `make fix` and commiting if there are changes...

File: calico/reference/resources/bgpconfig.md [hunk 2]
```diff
@@ -49,20 +49,21 @@ spec:
...
+| ignoredInterfaces      | List of network interfaces to be excluded when reading device routes.                                                                                                                                                                                                                                                                                                                                                          | A list of network interface names.                      | List of string                               | `nil` (no extra interfaces to be ignored) |
```
Comment (coutinhop) [inline calico/reference/resources/bgpconfig.md:66]: From looking at the code I got the impression that we can use wildcards, e.g. "interface-name*", is that correct? If so shouldn't we make it clearer that it is allowed in the documentation?

File: confd/pkg/backends/calico/client.go [hunk 3]
```diff
@@ -1349,6 +1350,15 @@ func (c *client) getNodeMeshPasswordKVPair(v3res *apiv3.BGPConfiguration, key in
 	}
 }
 
+func (c *client) getIgnoredInterfacesKVPair(v3res *apiv3.BGPConfiguration, key interface{}) {
+	ignoredInterfaces := getBGPConfigKey("ignored_interfaces", key)
+	if v3res != nil && v3res.Spec.IgnoredInterfaces != nil {
```
Comment (mgleung) [inline confd/pkg/backends/calico/client.go:1355]: nit: I don't expect this to happen, but I think we should also probably check if `v3rest.Spec` is `nil` too.

File: confd/pkg/backends/calico/client.go [hunk 4]
```diff
@@ -1349,6 +1350,15 @@ func (c *client) getNodeMeshPasswordKVPair(v3res *apiv3.BGPConfiguration, key in
 	}
 }
 
+func (c *client) getIgnoredInterfacesKVPair(v3res *apiv3.BGPConfiguration, key interface{}) {
+	ignoredInterfaces := getBGPConfigKey("ignored_interfaces", key)
```
Comment (mgleung) [inline confd/pkg/backends/calico/client.go:1354]: nit: might make this easier to follow if the variable was something like `ignoredIfacesKey` or `ignoredInterfacesKey`

File: confd/pkg/backends/calico/routes_test.go [hunk 5]
```diff
@@ -626,3 +630,21 @@ var _ = Describe("RouteGenerator", func() {
 		})
 	})
 })
+
+var _ = Describe("Update BGP Config Cache", func() {
+	c := &client{cache: make(map[string]string)}
+
+	It("should update cache value when IgnoredInterfaces set in BGPConfiguration", func() {
```
Comment (mgleung) [inline confd/pkg/backends/calico/routes_test.go:637]: nit: "should update cache value when IgnoredInterfaces is set in BGPConfiguration"

**Review by mgleung — APPROVED** [review]

**Pull request merged**